src/main/kotlin/com/londogard/textgen/smoothing/NoSmoothing.kt

```diff
@@ -0,0 +1,32 @@
+package com.londogard.textgen.smoothing
+
+import com.londogard.textgen.languagemodels.LanguageModel
+import com.londogard.textgen.languagemodels.LanguageModel.Companion.retrieveNgramData
+import com.londogard.textgen.normalization.Normalization
+import com.londogard.textgen.normalization.SimpleNormalization
+import com.londogard.textgen.penalties.Penalty
+import com.londogard.textgen.smoothing.Smoothing.Companion.takeP
+
+class NoSmoothing(
+    override val normalizer: Normalization = SimpleNormalization(),
+    override val penalties: List<Penalty> = emptyList()
+) : Smoothing {
+    override fun predict(languageModel: LanguageModel, history: List<Int>, token: Int): Double =
+        languageModel.internalLanguageModel[emptyList()]?.find { (key,_) -> key == token }?.second ?: 0.0
+
+    override fun probabilitiesTopK(
+        languageModel: LanguageModel,
+        history: List<Int>,
+        k: Int
+    ): List<Pair<Int, Double>> = languageModel.retrieveNgramData(history, penalties, languageModel.n)
+        .take(k)
```

Comment:
do we not have to sort the list?

Reply:
It's pre-sorted <3 

More expensive 'training' but cheaper 'inference' or 'usage'. :+1: